



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Sequence alignment tools">
      
      
        <link rel="canonical" href="https://biojulia.github.io/BioAlignments.jl/references/">
      
      
        <meta name="author" content="The BioJulia Organisation">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.3, mkdocs-material-2.7.3">
    
    
      
        <title>References - BioAlignments.jl</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.8d40d89b.css">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../assets/Documenter.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <a href="#references" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://biojulia.github.io/BioAlignments.jl/" title="BioAlignments.jl" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                BioAlignments.jl
              </span>
              <span class="md-header-nav__topic">
                References
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/BioJulia/BioAlignments.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <span class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </span>
    BioAlignments.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/BioJulia/BioAlignments.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../alignments/" title="Alignment representation" class="md-nav__link">
      Alignment representation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../pairalign/" title="Pairwise alignment" class="md-nav__link">
      Pairwise alignment
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../hts-files/" title="High-throughput sequencing" class="md-nav__link">
      High-throughput sequencing
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        References
      </label>
    
    <a href="./" title="References" class="md-nav__link md-nav__link--active">
      References
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#operations" title="Operations" class="md-nav__link">
    Operations
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignments" title="Alignments" class="md-nav__link">
    Alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#substitution-matrices" title="Substitution matrices" class="md-nav__link">
    Substitution matrices
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pairwise-alignments" title="Pairwise alignments" class="md-nav__link">
    Pairwise alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#io" title="I/O" class="md-nav__link">
    I/O
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sam" title="SAM" class="md-nav__link">
    SAM
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bam" title="BAM" class="md-nav__link">
    BAM
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../contributing/" title="Contributing" class="md-nav__link">
      Contributing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#operations" title="Operations" class="md-nav__link">
    Operations
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alignments" title="Alignments" class="md-nav__link">
    Alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#substitution-matrices" title="Substitution matrices" class="md-nav__link">
    Substitution matrices
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pairwise-alignments" title="Pairwise alignments" class="md-nav__link">
    Pairwise alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#io" title="I/O" class="md-nav__link">
    I/O
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sam" title="SAM" class="md-nav__link">
    SAM
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bam" title="BAM" class="md-nav__link">
    BAM
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/BioJulia/BioAlignments.jl/edit/master/docs/references.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='References-1'></a></p>
<h1 id="references">References</h1>
<p><a id='Operations-1'></a></p>
<h2 id="operations">Operations</h2>
<p><a id='BioAlignments.Operation' href='#BioAlignments.Operation'>#</a>
<strong><code>BioAlignments.Operation</code></strong> &mdash; <em>Type</em>.</p>
<p>Alignment operation.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L9-L11' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_MATCH' href='#BioAlignments.OP_MATCH'>#</a>
<strong><code>BioAlignments.OP_MATCH</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'M'</code>: non-specific match</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_INSERT' href='#BioAlignments.OP_INSERT'>#</a>
<strong><code>BioAlignments.OP_INSERT</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'I'</code>: insertion into reference sequence</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_DELETE' href='#BioAlignments.OP_DELETE'>#</a>
<strong><code>BioAlignments.OP_DELETE</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'D'</code>: deletion from reference sequence</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_SKIP' href='#BioAlignments.OP_SKIP'>#</a>
<strong><code>BioAlignments.OP_SKIP</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'N'</code>: (typically long) deletion from the reference, e.g. due to RNA splicing</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_SOFT_CLIP' href='#BioAlignments.OP_SOFT_CLIP'>#</a>
<strong><code>BioAlignments.OP_SOFT_CLIP</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'S'</code>: sequence removed from the beginning or end of the query sequence but stored</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_HARD_CLIP' href='#BioAlignments.OP_HARD_CLIP'>#</a>
<strong><code>BioAlignments.OP_HARD_CLIP</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'H'</code>: sequence removed from the beginning or end of the query sequence and not stored</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_PAD' href='#BioAlignments.OP_PAD'>#</a>
<strong><code>BioAlignments.OP_PAD</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'P'</code>: not currently supported, but present for SAM/BAM compatibility</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_SEQ_MATCH' href='#BioAlignments.OP_SEQ_MATCH'>#</a>
<strong><code>BioAlignments.OP_SEQ_MATCH</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'='</code>: match operation with matching sequence positions</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_SEQ_MISMATCH' href='#BioAlignments.OP_SEQ_MISMATCH'>#</a>
<strong><code>BioAlignments.OP_SEQ_MISMATCH</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'X'</code>: match operation with mismatching sequence positions</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_BACK' href='#BioAlignments.OP_BACK'>#</a>
<strong><code>BioAlignments.OP_BACK</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'B'</code>: not currently supported, but present for SAM/BAM compatibility</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OP_START' href='#BioAlignments.OP_START'>#</a>
<strong><code>BioAlignments.OP_START</code></strong> &mdash; <em>Constant</em>.</p>
<p><code>'0'</code>: indicate the start of an alignment within the reference and query sequence</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L41' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.ismatchop' href='#BioAlignments.ismatchop'>#</a>
<strong><code>BioAlignments.ismatchop</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ismatchop(op::Operation)
</pre></div>


<p>Test if <code>op</code> is a match operation (i.e. <code>op ∈ (OP_MATCH, OP_SEQ_MATCH, OP_SEQ_MISMATCH)</code>).</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L68-L72' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.isinsertop' href='#BioAlignments.isinsertop'>#</a>
<strong><code>BioAlignments.isinsertop</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isinsertop(op::Operation)
</pre></div>


<p>Test if <code>op</code> is a insertion operation (i.e. <code>op ∈ (OP_INSERT, OP_SOFT_CLIP, OP_HARD_CLIP)</code>).</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L77-L81' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.isdeleteop' href='#BioAlignments.isdeleteop'>#</a>
<strong><code>BioAlignments.isdeleteop</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isdeleteop(op::Operation)
</pre></div>


<p>Test if <code>op</code> is a deletion operation (i.e. <code>op ∈ (OP_DELETE, OP_SKIP)</code>).</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/operations.jl#L86-L90' class='documenter-source'>source</a><br></p>
<p><a id='Alignments-1'></a></p>
<h2 id="alignments">Alignments</h2>
<p><a id='BioAlignments.AlignmentAnchor' href='#BioAlignments.AlignmentAnchor'>#</a>
<strong><code>BioAlignments.AlignmentAnchor</code></strong> &mdash; <em>Type</em>.</p>
<p>Alignment operation with anchoring positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/anchors.jl#L9-L11' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.Alignment' href='#BioAlignments.Alignment'>#</a>
<strong><code>BioAlignments.Alignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Alignment of two sequences.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L9-L11' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.Alignment-Tuple{Array{BioAlignments.AlignmentAnchor,1},Bool}' href='#BioAlignments.Alignment-Tuple{Array{BioAlignments.AlignmentAnchor,1},Bool}'>#</a>
<strong><code>BioAlignments.Alignment</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>Alignment(anchors::Vector{AlignmentAnchor}, check=true)
</pre></div>


<p>Create an alignment object from a sequence of alignment anchors.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L9-L13' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.Alignment-Tuple{AbstractString,Int64,Int64}' href='#BioAlignments.Alignment-Tuple{AbstractString,Int64,Int64}'>#</a>
<strong><code>BioAlignments.Alignment</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>Alignment(cigar::AbstractString, seqpos=1, refpos=1)
</pre></div>


<p>Make an alignment object from a CIGAR string.</p>
<p><code>seqpos</code> and <code>refpos</code> specify the starting positions of two sequences.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L50-L56' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.seq2ref-Tuple{BioAlignments.Alignment,Integer}' href='#BioAlignments.seq2ref-Tuple{BioAlignments.Alignment,Integer}'>#</a>
<strong><code>BioAlignments.seq2ref</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>seq2ref(aln::Alignment, i::Integer)::Tuple{Int,Operation}
</pre></div>


<p>Map a position <code>i</code> from sequence to reference.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L103-L107' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.ref2seq-Tuple{BioAlignments.Alignment,Integer}' href='#BioAlignments.ref2seq-Tuple{BioAlignments.Alignment,Integer}'>#</a>
<strong><code>BioAlignments.ref2seq</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>ref2seq(aln::Alignment, i::Integer)::Tuple{Int,Operation}
</pre></div>


<p>Map a position <code>i</code> from reference to sequence.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L121-L125' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.cigar-Tuple{BioAlignments.Alignment}' href='#BioAlignments.cigar-Tuple{BioAlignments.Alignment}'>#</a>
<strong><code>BioAlignments.cigar</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>cigar(aln::Alignment)
</pre></div>


<p>Make a CIGAR string encoding of <code>aln</code>.</p>
<p>This is not entirely lossless as it discards the alignments start positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/alignment.jl#L139-L145' class='documenter-source'>source</a><br></p>
<p><a id='Substitution-matrices-1'></a></p>
<h2 id="substitution-matrices">Substitution matrices</h2>
<p><a id='BioAlignments.AbstractSubstitutionMatrix' href='#BioAlignments.AbstractSubstitutionMatrix'>#</a>
<strong><code>BioAlignments.AbstractSubstitutionMatrix</code></strong> &mdash; <em>Type</em>.</p>
<p>Supertype of substitution matrix.</p>
<p>The required method:</p>
<ul>
<li><code>Base.getindex(submat, x, y)</code>: substitution score/cost from <code>x</code> to <code>y</code></li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L9-L15' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SubstitutionMatrix' href='#BioAlignments.SubstitutionMatrix'>#</a>
<strong><code>BioAlignments.SubstitutionMatrix</code></strong> &mdash; <em>Type</em>.</p>
<p>Substitution matrix.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L18-L20' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.DichotomousSubstitutionMatrix' href='#BioAlignments.DichotomousSubstitutionMatrix'>#</a>
<strong><code>BioAlignments.DichotomousSubstitutionMatrix</code></strong> &mdash; <em>Type</em>.</p>
<p>Dichotomous substitution matrix.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L148-L150' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.EDNAFULL' href='#BioAlignments.EDNAFULL'>#</a>
<strong><code>BioAlignments.EDNAFULL</code></strong> &mdash; <em>Constant</em>.</p>
<p>EDNAFULL (or NUC4.4) substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L219' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.PAM30' href='#BioAlignments.PAM30'>#</a>
<strong><code>BioAlignments.PAM30</code></strong> &mdash; <em>Constant</em>.</p>
<p>PAM30 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L222' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.PAM70' href='#BioAlignments.PAM70'>#</a>
<strong><code>BioAlignments.PAM70</code></strong> &mdash; <em>Constant</em>.</p>
<p>PAM70 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L225' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.PAM250' href='#BioAlignments.PAM250'>#</a>
<strong><code>BioAlignments.PAM250</code></strong> &mdash; <em>Constant</em>.</p>
<p>PAM250 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L228' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BLOSUM45' href='#BioAlignments.BLOSUM45'>#</a>
<strong><code>BioAlignments.BLOSUM45</code></strong> &mdash; <em>Constant</em>.</p>
<p>BLOSUM45 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L231' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BLOSUM50' href='#BioAlignments.BLOSUM50'>#</a>
<strong><code>BioAlignments.BLOSUM50</code></strong> &mdash; <em>Constant</em>.</p>
<p>BLOSUM50 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L234' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BLOSUM62' href='#BioAlignments.BLOSUM62'>#</a>
<strong><code>BioAlignments.BLOSUM62</code></strong> &mdash; <em>Constant</em>.</p>
<p>BLOSUM62 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L237' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BLOSUM80' href='#BioAlignments.BLOSUM80'>#</a>
<strong><code>BioAlignments.BLOSUM80</code></strong> &mdash; <em>Constant</em>.</p>
<p>BLOSUM80 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L240' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BLOSUM90' href='#BioAlignments.BLOSUM90'>#</a>
<strong><code>BioAlignments.BLOSUM90</code></strong> &mdash; <em>Constant</em>.</p>
<p>BLOSUM90 substitution matrix</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/submat.jl#L243' class='documenter-source'>source</a><br></p>
<p><a id='Pairwise-alignments-1'></a></p>
<h2 id="pairwise-alignments">Pairwise alignments</h2>
<p><a id='BioAlignments.PairwiseAlignment' href='#BioAlignments.PairwiseAlignment'>#</a>
<strong><code>BioAlignments.PairwiseAlignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Pairwise alignment</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L9-L11' class='documenter-source'>source</a><br></p>
<p><a id='Base.count-Tuple{BioAlignments.PairwiseAlignment,BioAlignments.Operation}' href='#Base.count-Tuple{BioAlignments.PairwiseAlignment,BioAlignments.Operation}'>#</a>
<strong><code>Base.count</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>count(aln::PairwiseAlignment, target::Operation)
</pre></div>


<p>Count the number of positions where the <code>target</code> operation is applied.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L63-L67' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.count_matches' href='#BioAlignments.count_matches'>#</a>
<strong><code>BioAlignments.count_matches</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_matches(aln)
</pre></div>


<p>Count the number of matching positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L84-L88' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.count_mismatches' href='#BioAlignments.count_mismatches'>#</a>
<strong><code>BioAlignments.count_mismatches</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_mismatches(aln)
</pre></div>


<p>Count the number of mismatching positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L91-L95' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.count_insertions' href='#BioAlignments.count_insertions'>#</a>
<strong><code>BioAlignments.count_insertions</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_insertions(aln)
</pre></div>


<p>Count the number of inserting positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L98-L102' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.count_deletions' href='#BioAlignments.count_deletions'>#</a>
<strong><code>BioAlignments.count_deletions</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_deletions(aln)
</pre></div>


<p>Count the number of deleting positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L105-L109' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.count_aligned' href='#BioAlignments.count_aligned'>#</a>
<strong><code>BioAlignments.count_aligned</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>count_aligned(aln)
</pre></div>


<p>Count the number of aligned positions.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L112-L116' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.GlobalAlignment' href='#BioAlignments.GlobalAlignment'>#</a>
<strong><code>BioAlignments.GlobalAlignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Global-global alignment with end gap penalties.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L15' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SemiGlobalAlignment' href='#BioAlignments.SemiGlobalAlignment'>#</a>
<strong><code>BioAlignments.SemiGlobalAlignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Global-local alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L18' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.OverlapAlignment' href='#BioAlignments.OverlapAlignment'>#</a>
<strong><code>BioAlignments.OverlapAlignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Global-global alignment without end gap penalties.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L21' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.LocalAlignment' href='#BioAlignments.LocalAlignment'>#</a>
<strong><code>BioAlignments.LocalAlignment</code></strong> &mdash; <em>Type</em>.</p>
<p>Local-local alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L24' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.EditDistance' href='#BioAlignments.EditDistance'>#</a>
<strong><code>BioAlignments.EditDistance</code></strong> &mdash; <em>Type</em>.</p>
<p>Edit distance.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L31' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.HammingDistance' href='#BioAlignments.HammingDistance'>#</a>
<strong><code>BioAlignments.HammingDistance</code></strong> &mdash; <em>Type</em>.</p>
<p>Hamming distance.</p>
<p>A special case of <code>EditDistance</code> with the costs of insertion and deletion are infinitely large.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L42-L47' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.LevenshteinDistance' href='#BioAlignments.LevenshteinDistance'>#</a>
<strong><code>BioAlignments.LevenshteinDistance</code></strong> &mdash; <em>Type</em>.</p>
<p>Levenshtein distance.</p>
<p>A special case of <code>EditDistance</code> with the costs of mismatch, insertion, and deletion are 1.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/types.jl#L34-L39' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.AbstractScoreModel' href='#BioAlignments.AbstractScoreModel'>#</a>
<strong><code>BioAlignments.AbstractScoreModel</code></strong> &mdash; <em>Type</em>.</p>
<p>Supertype of score model.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/models.jl#L13-L15' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.AffineGapScoreModel' href='#BioAlignments.AffineGapScoreModel'>#</a>
<strong><code>BioAlignments.AffineGapScoreModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>AffineGapScoreModel(submat, gap_open, gap_extend)
AffineGapScoreModel(submat, gap_open=, gap_extend=)
AffineGapScoreModel(match=, mismatch=, gap_open=, gap_extend=)
</pre></div>


<p>Affine gap scoring model.</p>
<p>This creates an affine gap scroing model object for alignment from a substitution matrix (<code>submat</code>), a gap opening score (<code>gap_open</code>), and a gap extending score (<code>gap_extend</code>). A consecutive gap of length <code>k</code> has a score of <code>gap_open + gap_extend * k</code>. Note that both of the gap scores should be non-positive.  As a shorthand of creating a dichotomous substitution matrix, you can write as, for example, <code>AffineGapScoreModel(match=5, mismatch=-3, gap_open=-2, gap_extend=-1)</code>.</p>
<p><strong>Example</strong></p>
<div class="codehilite"><pre><span></span>using BioSequences
using BioAlignments

# create an affine gap scoring model from a predefined substitution
# matrix and gap opening/extending scores.
affinegap = AffineGapScoreModel(BLOSUM62, gap_open=-10, gap_extend=-1)

# run global alignment between two amino acid sequenecs
pairalign(GlobalAlignment(), aa&quot;IDGAAGQQL&quot;, aa&quot;IDGATGQL&quot;, affinegap)
</pre></div>


<p>See also: <code>SubstitutionMatrix</code>, <code>pairalign</code>, <code>CostModel</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/models.jl#L18-L47' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.AbstractCostModel' href='#BioAlignments.AbstractCostModel'>#</a>
<strong><code>BioAlignments.AbstractCostModel</code></strong> &mdash; <em>Type</em>.</p>
<p>Supertype of cost model.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/models.jl#L123-L125' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.CostModel' href='#BioAlignments.CostModel'>#</a>
<strong><code>BioAlignments.CostModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>CostModel(submat, insertion, deletion)
CostModel(submat, insertion=, deletion=)
CostModel(match=, mismatch=, insertion=, deletion=)
</pre></div>


<p>Cost model.</p>
<p>This creates a cost model object for alignment from substitution matrix (<code>submat</code>), an insertion cost (<code>insertion</code>), and a deletion cost (<code>deletion</code>). Note that both of the insertion and deletion costs should be non-negative.  As a shorthand of creating a dichotomous substitution matrix, you can write as, for example, <code>CostModel(match=0, mismatch=1, insertion=2, deletion=2)</code>.</p>
<p><strong>Example</strong></p>
<div class="codehilite"><pre><span></span>using BioAlignments

# create a cost model from a substitution matrix and indel costs
cost = CostModel(ones(128, 128) - eye(128), insertion=.5, deletion=.5)

# run global alignment to minimize edit distance
pairalign(EditDistance(), &quot;intension&quot;, &quot;execution&quot;, cost)
</pre></div>


<p>See also: <code>SubstitutionMatrix</code>, <code>pairalign</code>, <code>AffineGapScoreModel</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/models.jl#L128-L153' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.PairwiseAlignmentResult' href='#BioAlignments.PairwiseAlignmentResult'>#</a>
<strong><code>BioAlignments.PairwiseAlignmentResult</code></strong> &mdash; <em>Type</em>.</p>
<p>Result of pairwise alignment</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/result.jl#L9-L11' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.pairalign' href='#BioAlignments.pairalign'>#</a>
<strong><code>BioAlignments.pairalign</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>pairalign(type, seq, ref, model, [options...])
</pre></div>


<p>Run pairwise alignment between two sequences: <code>seq</code> and <code>ref</code>.</p>
<p>Available <code>type</code>s are:</p>
<ul>
<li><code>GlobalAlignment()</code></li>
<li><code>LocalAlignment()</code></li>
<li><code>SemiGlobalAlignment()</code></li>
<li><code>OverlapAlignment()</code></li>
<li><code>EditDistance()</code></li>
<li><code>LevenshteinDistance()</code></li>
<li><code>HammingDistance()</code></li>
</ul>
<p><code>GlobalAlignment</code>, <code>LocalAlignment</code>, <code>SemiGlobalAlignment</code>, and <code>OverlapAlignment</code> are problem that maximizes alignment score between two sequences.  Therefore, <code>model</code> should be an instance of <code>AbstractScoreModel</code> (e.g. <code>AffineGapScoreModel</code>).</p>
<p><code>EditDistance</code>, <code>LevenshteinDistance</code>, and <code>HammingDistance</code> are problem that minimizes alignment cost between two sequences.  As for <code>EditDistance</code>, <code>model</code> should be an instance of <code>AbstractCostModel</code> (e.g. <code>CostModel</code>). <code>LevenshteinDistance</code> and <code>HammingDistance</code> have predefined a cost model, so users cannot specify a cost model for these alignment types.</p>
<p>When you pass the <code>score_only=true</code> or <code>distance_only=true</code> option to <code>pairalign</code>, the result of pairwise alignment holds alignment score/distance only.  This may enable some algorithms to run faster than calculating full alignment result.  Other available <code>options</code> are documented for each alignemnt type.</p>
<p><strong>Example</strong></p>
<div class="codehilite"><pre><span></span>using BioSequences
using BioAlignments

# create affine gap scoring model
affinegap = AffineGapScoreModel(
    match=5,
    mismatch=-4,
    gap_open=-5,
    gap_extend=-3
)

# run global alignment between two DNA sequences
pairalign(GlobalAlignment(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;, affinegap)

# run local alignment between two DNA sequences
pairalign(LocalAlignment(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;, affinegap)

# you cannot specify a cost model in LevenshteinDistance
pairalign(LevenshteinDistance(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;)
</pre></div>


<p>See also: <code>AffineGapScoreModel</code>, <code>CostModel</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/pairalign.jl#L19-L75' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.score' href='#BioAlignments.score'>#</a>
<strong><code>BioAlignments.score</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>score(alignment_result)
</pre></div>


<p>Return score of alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/result.jl#L33-L37' class='documenter-source'>source</a><br></p>
<p><a id='BioCore.distance' href='#BioCore.distance'>#</a>
<strong><code>BioCore.distance</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>distance(alignment_result)
</pre></div>


<p>Retrun distance of alignment.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/result.jl#L40-L44' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.alignment' href='#BioAlignments.alignment'>#</a>
<strong><code>BioAlignments.alignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignment(alignment_result)
</pre></div>


<p>Return alignment if any.</p>
<p>See also: <code>hasalignment</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/result.jl#L55-L61' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.hasalignment' href='#BioAlignments.hasalignment'>#</a>
<strong><code>BioAlignments.hasalignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>hasalignment(alignment_result)
</pre></div>


<p>Check if alignment is stored or not.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/result.jl#L48-L52' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.seq2ref-Tuple{BioAlignments.PairwiseAlignment,Integer}' href='#BioAlignments.seq2ref-Tuple{BioAlignments.PairwiseAlignment,Integer}'>#</a>
<strong><code>BioAlignments.seq2ref</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>seq2ref(aln::PairwiseAlignment, i::Integer)::Tuple{Int,Operation}
</pre></div>


<p>Map a position <code>i</code> from the first sequence to the second.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L131-L135' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.ref2seq-Tuple{BioAlignments.PairwiseAlignment,Integer}' href='#BioAlignments.ref2seq-Tuple{BioAlignments.PairwiseAlignment,Integer}'>#</a>
<strong><code>BioAlignments.ref2seq</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>ref2seq(aln::PairwiseAlignment, i::Integer)::Tuple{Int,Operation}
</pre></div>


<p>Map a position <code>i</code> from the second sequence to the first.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/pairwise/alignment.jl#L140-L144' class='documenter-source'>source</a><br></p>
<p><a id='I/O-1'></a></p>
<h2 id="io">I/O</h2>
<p><a id='SAM-1'></a></p>
<h3 id="sam">SAM</h3>
<p><a id='BioAlignments.SAM.Reader' href='#BioAlignments.SAM.Reader'>#</a>
<strong><code>BioAlignments.SAM.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>SAM.Reader(input::IO)
</pre></div>


<p>Create a data reader of the SAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/reader.jl#L16-L23' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.header' href='#BioAlignments.SAM.header'>#</a>
<strong><code>BioAlignments.SAM.header</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>header(reader::Reader)::Header
</pre></div>


<p>Get the header of <code>reader</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/reader.jl#L32-L36' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.Header' href='#BioAlignments.SAM.Header'>#</a>
<strong><code>BioAlignments.SAM.Header</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>SAM.Header()
</pre></div>


<p>Create an empty header.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/header.jl#L8-L12' class='documenter-source'>source</a><br></p>
<p><a id='Base.find-Tuple{BioAlignments.SAM.Header,AbstractString}' href='#Base.find-Tuple{BioAlignments.SAM.Header,AbstractString}'>#</a>
<strong><code>Base.find</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>find(header::Header, key::AbstractString)::Vector{MetaInfo}
</pre></div>


<p>Find metainfo objects satisfying <code>SAM.tag(metainfo) == key</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/header.jl#L41-L45' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.Writer' href='#BioAlignments.SAM.Writer'>#</a>
<strong><code>BioAlignments.SAM.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>Writer(output::IO, header::Header=Header())
</pre></div>


<p>Create a data writer of the SAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header=Header()</code>: SAM header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/writer.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.MetaInfo' href='#BioAlignments.SAM.MetaInfo'>#</a>
<strong><code>BioAlignments.SAM.MetaInfo</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>MetaInfo(str::AbstractString)
</pre></div>


<p>Create a SAM metainfo from <code>str</code>.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; SAM.MetaInfo(&quot;@CO    some comment&quot;)
BioAlignments.SAM.MetaInfo:
    tag: CO
  value: some comment

julia&gt; SAM.MetaInfo(&quot;@SQ    SN:chr1 LN:12345&quot;)
BioAlignments.SAM.MetaInfo:
    tag: SQ
  value: SN=chr1 LN=12345
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L23-L40' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>MetaInfo(tag::AbstractString, value)
</pre></div>


<p>Create a SAM metainfo with <code>tag</code> and <code>value</code>.</p>
<p><code>tag</code> is a two-byte ASCII string. If <code>tag</code> is <code>"CO"</code>, <code>value</code> must be a string; otherwise, <code>value</code> is an iterable object with key and value pairs.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; SAM.MetaInfo(&quot;CO&quot;, &quot;some comment&quot;)
BioAlignments.SAM.MetaInfo:
    tag: CO
  value: some comment

julia&gt; string(ans)
&quot;@CO    some comment&quot;

julia&gt; SAM.MetaInfo(&quot;SQ&quot;, [&quot;SN&quot; =&gt; &quot;chr1&quot;, &quot;LN&quot; =&gt; 12345])
BioAlignments.SAM.MetaInfo:
    tag: SQ
  value: SN=chr1 LN=12345

julia&gt; string(ans)
&quot;@SQ    SN:chr1 LN:12345&quot;
</pre></div>


<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L45-L71' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.iscomment' href='#BioAlignments.SAM.iscomment'>#</a>
<strong><code>BioAlignments.SAM.iscomment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>iscomment(metainfo::MetaInfo)::Bool
</pre></div>


<p>Test if <code>metainfo</code> is a comment (i.e. its tag is "CO").</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L154-L158' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.tag' href='#BioAlignments.SAM.tag'>#</a>
<strong><code>BioAlignments.SAM.tag</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>tag(metainfo::MetaInfo)::String
</pre></div>


<p>Get the tag of <code>metainfo</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L163-L167' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.value' href='#BioAlignments.SAM.value'>#</a>
<strong><code>BioAlignments.SAM.value</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>value(metainfo::MetaInfo)::String
</pre></div>


<p>Get the value of <code>metainfo</code> as a string.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L173-L177' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.keyvalues' href='#BioAlignments.SAM.keyvalues'>#</a>
<strong><code>BioAlignments.SAM.keyvalues</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>keyvalues(metainfo::MetaInfo)::Vector{Pair{String,String}}
</pre></div>


<p>Get the values of <code>metainfo</code> as string pairs.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/metainfo.jl#L183-L187' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.Record' href='#BioAlignments.SAM.Record'>#</a>
<strong><code>BioAlignments.SAM.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>SAM.Record()
</pre></div>


<p>Create an unfilled SAM record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L23-L27' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>SAM.Record(data::Vector{UInt8})
</pre></div>


<p>Create a SAM record from <code>data</code>. This function verifies the format and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L39-L45' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>SAM.Record(str::AbstractString)
</pre></div>


<p>Create a SAM record from <code>str</code>. This function verifies the format and indexes fields for accessors.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L63-L68' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.flag' href='#BioAlignments.SAM.flag'>#</a>
<strong><code>BioAlignments.SAM.flag</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>flag(record::Record)::UInt16
</pre></div>


<p>Get the bitwise flag of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L133-L137' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.ismapped' href='#BioAlignments.SAM.ismapped'>#</a>
<strong><code>BioAlignments.SAM.ismapped</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ismapped(record::Record)::Bool
</pre></div>


<p>Test if <code>record</code> is mapped.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L147-L151' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.isprimary' href='#BioAlignments.SAM.isprimary'>#</a>
<strong><code>BioAlignments.SAM.isprimary</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isprimary(record::Record)::Bool
</pre></div>


<p>Test if <code>record</code> is a primary line of the read.</p>
<p>This is equivalent to <code>flag(record) &amp; 0x900 == 0</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L156-L162' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.refname' href='#BioAlignments.SAM.refname'>#</a>
<strong><code>BioAlignments.SAM.refname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>refname(record::Record)::String
</pre></div>


<p>Get the reference sequence name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L167-L171' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.position' href='#BioAlignments.SAM.position'>#</a>
<strong><code>BioAlignments.SAM.position</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>position(record::Record)::Int
</pre></div>


<p>Get the 1-based leftmost mapping position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L184-L188' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.rightposition' href='#BioAlignments.SAM.rightposition'>#</a>
<strong><code>BioAlignments.SAM.rightposition</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>rightposition(record::Record)::Int
</pre></div>


<p>Get the 1-based rightmost mapping position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L202-L206' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.isnextmapped' href='#BioAlignments.SAM.isnextmapped'>#</a>
<strong><code>BioAlignments.SAM.isnextmapped</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isnextmapped(record::Record)::Bool
</pre></div>


<p>Test if the mate/next read of <code>record</code> is mapped.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L215-L219' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.nextrefname' href='#BioAlignments.SAM.nextrefname'>#</a>
<strong><code>BioAlignments.SAM.nextrefname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nextrefname(record::Record)::String
</pre></div>


<p>Get the reference name of the mate/next read of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L224-L228' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.nextposition' href='#BioAlignments.SAM.nextposition'>#</a>
<strong><code>BioAlignments.SAM.nextposition</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nextposition(record::Record)::Int
</pre></div>


<p>Get the position of the mate/next read of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L241-L245' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.mappingquality' href='#BioAlignments.SAM.mappingquality'>#</a>
<strong><code>BioAlignments.SAM.mappingquality</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>mappingquality(record::Record)::UInt8
</pre></div>


<p>Get the mapping quality of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L259-L263' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.cigar' href='#BioAlignments.SAM.cigar'>#</a>
<strong><code>BioAlignments.SAM.cigar</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>cigar(record::Record)::String
</pre></div>


<p>Get the CIGAR string of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L277-L281' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.alignment' href='#BioAlignments.SAM.alignment'>#</a>
<strong><code>BioAlignments.SAM.alignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignment(record::Record)::BioAlignments.Alignment
</pre></div>


<p>Get the alignment of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L294-L298' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.alignlength' href='#BioAlignments.SAM.alignlength'>#</a>
<strong><code>BioAlignments.SAM.alignlength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignlength(record::Record)::Int
</pre></div>


<p>Get the alignment length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L311-L315' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.tempname' href='#BioAlignments.SAM.tempname'>#</a>
<strong><code>BioAlignments.SAM.tempname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>tempname(record::Record)::String
</pre></div>


<p>Get the query template name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L337-L341' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.templength' href='#BioAlignments.SAM.templength'>#</a>
<strong><code>BioAlignments.SAM.templength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>templength(record::Record)::Int
</pre></div>


<p>Get the template length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L354-L358' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.sequence' href='#BioAlignments.SAM.sequence'>#</a>
<strong><code>BioAlignments.SAM.sequence</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sequence(record::Record)::BioSequences.DNASequence
</pre></div>


<p>Get the segment sequence of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L372-L376' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>sequence(::Type{String}, record::Record)::String
</pre></div>


<p>Get the segment sequence of <code>record</code> as <code>String</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L391-L395' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.seqlength' href='#BioAlignments.SAM.seqlength'>#</a>
<strong><code>BioAlignments.SAM.seqlength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seqlength(record::Record)::Int
</pre></div>


<p>Get the sequence length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L401-L405' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.quality' href='#BioAlignments.SAM.quality'>#</a>
<strong><code>BioAlignments.SAM.quality</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>quality(record::Record)::Vector{UInt8}
</pre></div>


<p>Get the Phred-scaled base quality of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L418-L422' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>quality(::Type{String}, record::Record)::String
</pre></div>


<p>Get the ASCII-encoded base quality of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L439-L443' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.auxdata' href='#BioAlignments.SAM.auxdata'>#</a>
<strong><code>BioAlignments.SAM.auxdata</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>auxdata(record::Record)::Dict{String,Any}
</pre></div>


<p>Get the auxiliary data (optional fields) of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/record.jl#L449-L453' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_PAIRED' href='#BioAlignments.SAM.FLAG_PAIRED'>#</a>
<strong><code>BioAlignments.SAM.FLAG_PAIRED</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0001: the read is paired in sequencing, no matter whether it is mapped in a pair</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_PROPER_PAIR' href='#BioAlignments.SAM.FLAG_PROPER_PAIR'>#</a>
<strong><code>BioAlignments.SAM.FLAG_PROPER_PAIR</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0002: the read is mapped in a proper pair</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_UNMAP' href='#BioAlignments.SAM.FLAG_UNMAP'>#</a>
<strong><code>BioAlignments.SAM.FLAG_UNMAP</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0004: the read itself is unmapped; conflictive with SAM.FLAG_PROPER_PAIR</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_MUNMAP' href='#BioAlignments.SAM.FLAG_MUNMAP'>#</a>
<strong><code>BioAlignments.SAM.FLAG_MUNMAP</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0008: the mate is unmapped</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_REVERSE' href='#BioAlignments.SAM.FLAG_REVERSE'>#</a>
<strong><code>BioAlignments.SAM.FLAG_REVERSE</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0010: the read is mapped to the reverse strand</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_MREVERSE' href='#BioAlignments.SAM.FLAG_MREVERSE'>#</a>
<strong><code>BioAlignments.SAM.FLAG_MREVERSE</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0020: the mate is mapped to the reverse strand</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_READ1' href='#BioAlignments.SAM.FLAG_READ1'>#</a>
<strong><code>BioAlignments.SAM.FLAG_READ1</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0040: this is read1</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_READ2' href='#BioAlignments.SAM.FLAG_READ2'>#</a>
<strong><code>BioAlignments.SAM.FLAG_READ2</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0080: this is read2</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_SECONDARY' href='#BioAlignments.SAM.FLAG_SECONDARY'>#</a>
<strong><code>BioAlignments.SAM.FLAG_SECONDARY</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0100: not primary alignment</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_QCFAIL' href='#BioAlignments.SAM.FLAG_QCFAIL'>#</a>
<strong><code>BioAlignments.SAM.FLAG_QCFAIL</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0200: QC failure</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_DUP' href='#BioAlignments.SAM.FLAG_DUP'>#</a>
<strong><code>BioAlignments.SAM.FLAG_DUP</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0400: optical or PCR duplicate</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.SAM.FLAG_SUPPLEMENTARY' href='#BioAlignments.SAM.FLAG_SUPPLEMENTARY'>#</a>
<strong><code>BioAlignments.SAM.FLAG_SUPPLEMENTARY</code></strong> &mdash; <em>Constant</em>.</p>
<p>0x0800: supplementary alignment</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/sam/flags.jl#L6' class='documenter-source'>source</a><br></p>
<p><a id='BAM-1'></a></p>
<h3 id="bam">BAM</h3>
<p><a id='BioAlignments.BAM.Reader' href='#BioAlignments.BAM.Reader'>#</a>
<strong><code>BioAlignments.BAM.Reader</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">BAM</span><span class="p">.</span><span class="nc">Reader</span><span class="o">(</span><span class="nt">input</span><span class="p">::</span><span class="nd">IO</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">nothing</span><span class="o">)</span>
</pre></div>


<p>Create a data reader of the BAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>input</code>: data source</li>
<li><code>index=nothing</code>: filepath to a random access index (currently <em>bai</em> is supported)</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/reader.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.header' href='#BioAlignments.BAM.header'>#</a>
<strong><code>BioAlignments.BAM.header</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">header</span><span class="o">(</span><span class="nt">reader</span><span class="p">::</span><span class="nd">Reader</span><span class="o">;</span> <span class="nt">fillSQ</span><span class="p">::</span><span class="nd">Bool</span><span class="o">=</span><span class="nt">false</span><span class="o">)</span><span class="p">::</span><span class="nd">SAM</span><span class="p">.</span><span class="nc">Header</span>
</pre></div>


<p>Get the header of <code>reader</code>.</p>
<p>If <code>fillSQ</code> is <code>true</code>, this function fills missing "SQ" metainfo in the header.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/reader.jl#L48-L54' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.Writer' href='#BioAlignments.BAM.Writer'>#</a>
<strong><code>BioAlignments.BAM.Writer</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BAM.Writer(output::BGZFStream, header::SAM.Header)
</pre></div>


<p>Create a data writer of the BAM file format.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>output</code>: data sink</li>
<li><code>header</code>: SAM header object</li>
</ul>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/writer.jl#L4-L12' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.Record' href='#BioAlignments.BAM.Record'>#</a>
<strong><code>BioAlignments.BAM.Record</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BAM.Record()
</pre></div>


<p>Create an unfilled BAM record.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L4-L8' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.flag' href='#BioAlignments.BAM.flag'>#</a>
<strong><code>BioAlignments.BAM.flag</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>flag(record::Record)::UInt16
</pre></div>


<p>Get the bitwise flag of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L96-L100' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.ismapped' href='#BioAlignments.BAM.ismapped'>#</a>
<strong><code>BioAlignments.BAM.ismapped</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>ismapped(record::Record)::Bool
</pre></div>


<p>Test if <code>record</code> is mapped.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L110-L114' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.isprimary' href='#BioAlignments.BAM.isprimary'>#</a>
<strong><code>BioAlignments.BAM.isprimary</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isprimary(record::Record)::Bool
</pre></div>


<p>Test if <code>record</code> is a primary line of the read.</p>
<p>This is equivalent to <code>flag(record) &amp; 0x900 == 0</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L119-L125' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.refid' href='#BioAlignments.BAM.refid'>#</a>
<strong><code>BioAlignments.BAM.refid</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>refid(record::Record)::Int
</pre></div>


<p>Get the reference sequence ID of <code>record</code>.</p>
<p>The ID is 1-based (i.e. the first sequence is 1) and is 0 for a record without a mapping position.</p>
<p>See also: <code>BAM.rname</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L130-L138' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.refname' href='#BioAlignments.BAM.refname'>#</a>
<strong><code>BioAlignments.BAM.refname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>refname(record::Record)::String
</pre></div>


<p>Get the reference sequence name of <code>record</code>.</p>
<p>See also: <code>BAM.refid</code></p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L148-L154' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.position' href='#BioAlignments.BAM.position'>#</a>
<strong><code>BioAlignments.BAM.position</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>position(record::Record)::Int
</pre></div>


<p>Get the 1-based leftmost mapping position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L170-L174' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.rightposition' href='#BioAlignments.BAM.rightposition'>#</a>
<strong><code>BioAlignments.BAM.rightposition</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>rightposition(record::Record)::Int
</pre></div>


<p>Get the 1-based rightmost mapping position of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L184-L188' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.isnextmapped' href='#BioAlignments.BAM.isnextmapped'>#</a>
<strong><code>BioAlignments.BAM.isnextmapped</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>isnextmapped(record::Record)::Bool
</pre></div>


<p>Test if the mate/next read of <code>record</code> is mapped.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L198-L202' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.nextrefid' href='#BioAlignments.BAM.nextrefid'>#</a>
<strong><code>BioAlignments.BAM.nextrefid</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nextrefid(record::Record)::Int
</pre></div>


<p>Get the next/mate reference sequence ID of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L207-L211' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.nextrefname' href='#BioAlignments.BAM.nextrefname'>#</a>
<strong><code>BioAlignments.BAM.nextrefname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nextrefname(record::Record)::String
</pre></div>


<p>Get the reference name of the mate/next read of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L221-L225' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.nextposition' href='#BioAlignments.BAM.nextposition'>#</a>
<strong><code>BioAlignments.BAM.nextposition</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nextposition(record::Record)::Int
</pre></div>


<p>Get the 1-based leftmost mapping position of the next/mate read of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L241-L245' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.mappingquality' href='#BioAlignments.BAM.mappingquality'>#</a>
<strong><code>BioAlignments.BAM.mappingquality</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>mappingquality(record::Record)::UInt8
</pre></div>


<p>Get the mapping quality of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L255-L259' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.cigar' href='#BioAlignments.BAM.cigar'>#</a>
<strong><code>BioAlignments.BAM.cigar</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>cigar(record::Record)::String
</pre></div>


<p>Get the CIGAR string of <code>record</code>.</p>
<p>See also <code>BAM.cigar_rle</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L268-L274' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.cigar_rle' href='#BioAlignments.BAM.cigar_rle'>#</a>
<strong><code>BioAlignments.BAM.cigar_rle</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>cigar_rle(record::Record)::Tuple{Vector{BioAlignments.Operation},Vector{Int}}
</pre></div>


<p>Get a run-length encoded tuple <code>(ops, lens)</code> of the CIGAR string in <code>record</code>.</p>
<p>See also <code>BAM.cigar</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L283-L289' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.alignment' href='#BioAlignments.BAM.alignment'>#</a>
<strong><code>BioAlignments.BAM.alignment</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignment(record::Record)::BioAlignments.Alignment
</pre></div>


<p>Get the alignment of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L304-L308' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.alignlength' href='#BioAlignments.BAM.alignlength'>#</a>
<strong><code>BioAlignments.BAM.alignlength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>alignlength(record::Record)::Int
</pre></div>


<p>Get the alignment length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L337-L341' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.tempname' href='#BioAlignments.BAM.tempname'>#</a>
<strong><code>BioAlignments.BAM.tempname</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>tempname(record::Record)::String
</pre></div>


<p>Get the query template name of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L355-L359' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.templength' href='#BioAlignments.BAM.templength'>#</a>
<strong><code>BioAlignments.BAM.templength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>templength(record::Record)::Int
</pre></div>


<p>Get the template length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L370-L374' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.sequence' href='#BioAlignments.BAM.sequence'>#</a>
<strong><code>BioAlignments.BAM.sequence</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sequence(record::Record)::BioSequences.DNASequence
</pre></div>


<p>Get the segment sequence of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L384-L388' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.seqlength' href='#BioAlignments.BAM.seqlength'>#</a>
<strong><code>BioAlignments.BAM.seqlength</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>seqlength(record::Record)::Int
</pre></div>


<p>Get the sequence length of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L406-L410' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.quality' href='#BioAlignments.BAM.quality'>#</a>
<strong><code>BioAlignments.BAM.quality</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>quality(record::Record)::Vector{UInt8}
</pre></div>


<p>Get the base quality of  <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L420-L424' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.auxdata' href='#BioAlignments.BAM.auxdata'>#</a>
<strong><code>BioAlignments.BAM.auxdata</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>auxdata(record::Record)::BAM.AuxData
</pre></div>


<p>Get the auxiliary data of <code>record</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/record.jl#L436-L440' class='documenter-source'>source</a><br></p>
<p><a id='BioAlignments.BAM.BAI' href='#BioAlignments.BAM.BAI'>#</a>
<strong><code>BioAlignments.BAM.BAI</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BAI(filename::AbstractString)
</pre></div>


<p>Load a BAI index from <code>filename</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/bai.jl#L18-L22' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>BAI(input::IO)
</pre></div>


<p>Load a BAI index from <code>input</code>.</p>
<p><a target='_blank' href='https://github.com/BioJulia/BioAlignments.jl/blob/11d47d7aad4d51270dc791d0f0a67f2cbdd00e1d/src/bam/bai.jl#L27-L31' class='documenter-source'>source</a><br></p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../hts-files/" title="High-throughput sequencing" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                High-throughput sequencing
              </span>
            </div>
          </a>
        
        
          <a href="../contributing/" title="Contributing" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Contributing
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright (c) 2017 The BioJulia Organisation
          </div>
        
        powered by
        <a href="http://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.b438e6c5.js"></script>
      
      <script>app.initialize({version:"0.17.3",url:{base:".."}})</script>
      
    
    
      
    
  </body>
</html>