
<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Sequence alignment tools">
      
      
        <link rel="canonical" href="https://biojulia.github.io/BioAlignments.jl/hts-files/">
      
      
        <meta name="author" content="The BioJulia Organisation">
      
      
        <link rel="shortcut icon" href="../assets/images/favicon.png">
      
      <meta name="generator" content="mkdocs-0.16.3, mkdocs-material-1.8.0">
    
    
      
        <title>High-throughput sequencing - BioAlignments.jl</title>
      
    
    
      <script src="../assets/javascripts/modernizr-1df76c4e58.js"></script>
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application-bfecc7305d.css">
      
    
    
      
        
        
        
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../assets/Documenter.css">
    
    
  </head>
  
  
  
  
    <body>
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <a href="https://biojulia.github.io/BioAlignments.jl/" title="BioAlignments.jl" class="md-icon md-icon--home md-header-nav__button">
          </a>
        
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <span class="md-flex__ellipsis md-header-nav__title">
          
            
              
            
            High-throughput sequencing
          
        </span>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" required placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset">close</button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result" data-md-lang-search="" data-md-lang-tokenizer="[\s\-]+">
          <div class="md-search-result__meta" data-md-lang-result-none="No matching documents" data-md-lang-result-one="1 matching document" data-md-lang-result-other="# matching documents">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/BioJulia/BioAlignments.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    
      <i class="md-icon md-icon--home md-nav__button"></i>
    
    BioAlignments.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/BioJulia/BioAlignments.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../alignments/" title="Alignment representation" class="md-nav__link">
      Alignment representation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../pairalign/" title="Pairwise alignment" class="md-nav__link">
      Pairwise alignment
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        High-throughput sequencing
      </label>
    
    <a href="./" title="High-throughput sequencing" class="md-nav__link md-nav__link--active">
      High-throughput sequencing
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" title="Overview" class="md-nav__link">
    Overview
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sam-and-bam-file-formats" title="SAM and BAM file formats" class="md-nav__link">
    SAM and BAM file formats
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-records-in-a-range" title="Getting records in a range" class="md-nav__link">
    Getting records in a range
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performance-tips" title="Performance tips" class="md-nav__link">
    Performance tips
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../references/" title="References" class="md-nav__link">
      References
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../contributing/" title="Contributing" class="md-nav__link">
      Contributing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" title="Overview" class="md-nav__link">
    Overview
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sam-and-bam-file-formats" title="SAM and BAM file formats" class="md-nav__link">
    SAM and BAM file formats
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-records-in-a-range" title="Getting records in a range" class="md-nav__link">
    Getting records in a range
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performance-tips" title="Performance tips" class="md-nav__link">
    Performance tips
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/BioJulia/BioAlignments.jl/edit/master/docs/hts-files.md" title="Edit this page" class="md-icon md-content__icon">edit</a>
                
                
                <p><a id='High-throughput-Sequencing-1'></a></p>
<h1 id="high-throughput-sequencing">High-throughput Sequencing</h1>
<p><a id='Overview-1'></a></p>
<h2 id="overview">Overview</h2>
<p>High-throughput sequencing (HTS) technologies generate a large amount of data in the form of a large number of nucleotide sequencing reads. One of the most common tasks in bioinformatics is to align these reads against known reference genomes, chromosomes, or contigs. BioAlignments.jl provides several data formats commonly used for this kind of task.</p>
<p><a id='SAM-and-BAM-file-formats-1'></a></p>
<h2 id="sam-and-bam-file-formats">SAM and BAM file formats</h2>
<p>BioAlignments.jl offers high-performance tools for SAM and BAM file formats, which are the most popular file formats.  A typical code iterating over all records in a file looks like below:</p>
<div class="codehilite"><pre><span></span><span class="c"># Import the SAM and BAM module.</span>
<span class="k">using</span> <span class="n">BioAlignments</span>

<span class="c"># Open a BAM file.</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>

<span class="c"># Iterate over BAM records.</span>
<span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">reader</span>
    <span class="c"># `record` is a BAM.Record object.</span>
    <span class="k">if</span> <span class="n">BAM</span><span class="o">.</span><span class="n">ismapped</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
        <span class="c"># Print the mapped position.</span>
        <span class="n">println</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">refname</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="sc">&#39;:&#39;</span><span class="p">,</span> <span class="n">BAM</span><span class="o">.</span><span class="n">position</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>
    <span class="k">end</span>
<span class="k">end</span>

<span class="c"># Close the BAM file.</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p>Accessor functions are defined in <code>SAM</code> and <code>BAM</code> modules.  Lists of these functions to <code>SAM.Record</code> and <code>BAM.Record</code> are described in the <a href="../references/#SAM-1">SAM</a> and <a href="../references/#BAM-1">BAM</a> sections, respectively.</p>
<p><code>SAM.Reader</code> and <code>BAM.Reader</code> implement the <code>header</code> function, which returns a <code>SAM.Header</code> object. This is conceptually a sequence of <code>SAM.MetaInfo</code> objects corresponding to header lines that start with '@' markers. To select <code>SAM.MetaInfo</code> records with a specific tag, you can use the <code>find</code> function:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">SAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;data.sam&quot;</span><span class="p">);</span>

<span class="gp">julia&gt;</span> <span class="n">find</span><span class="p">(</span><span class="n">header</span><span class="p">(</span><span class="n">reader</span><span class="p">),</span> <span class="s">&quot;SQ&quot;</span><span class="p">)</span>
<span class="go">7-element Array{Bio.Align.SAM.MetaInfo,1}:</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=Chr1 LN=30427671</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=Chr2 LN=19698289</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=Chr3 LN=23459830</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=Chr4 LN=18585056</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=Chr5 LN=26975502</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=chloroplast LN=154478</span>
<span class="go"> Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=mitochondria LN=366924</span>
</pre></div>


<p>A <code>SAM.MetaInfo</code> object can be created as follows:</p>
<div class="codehilite"><pre><span></span><span class="gp">julia&gt;</span> <span class="n">SAM</span><span class="o">.</span><span class="n">MetaInfo</span><span class="p">(</span><span class="s">&quot;SQ&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s">&quot;SN&quot;</span> <span class="o">=&gt;</span> <span class="s">&quot;chr1&quot;</span><span class="p">,</span> <span class="s">&quot;LN&quot;</span> <span class="o">=&gt;</span> <span class="mi">1234</span><span class="p">])</span>
<span class="go">Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: SQ</span>
<span class="go">  value: SN=chr1 LN=1234</span>

<span class="gp">julia&gt;</span> <span class="n">SAM</span><span class="o">.</span><span class="n">MetaInfo</span><span class="p">(</span><span class="s">&quot;CO&quot;</span><span class="p">,</span> <span class="s">&quot;comment&quot;</span><span class="p">)</span>
<span class="go">Bio.Align.SAM.MetaInfo:</span>
<span class="go">    tag: CO</span>
<span class="go">  value: comment</span>
</pre></div>


<p><a id='Getting-records-in-a-range-1'></a></p>
<h2 id="getting-records-in-a-range">Getting records in a range</h2>
<p>BioAlignments.jl supports the BAI index to fetch records in a specific range from a BAM file.  <a href="https://samtools.github.io/">Samtools</a> provides <code>index</code> subcommand to create an index file (.bai) from a sorted BAM file.</p>
<div class="codehilite"><pre><span></span><span class="gp">$</span> samtools index -b SRR1238088.sort.bam
<span class="gp">$</span> ls SRR1238088.sort.bam*
<span class="go">SRR1238088.sort.bam     SRR1238088.sort.bam.bai</span>
</pre></div>


<p><code>eachoverlap(reader, chrom, range)</code> returns an iterator of BAM records overlapping the query interval:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;SRR1238088.sort.bam&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s">&quot;SRR1238088.sort.bam.bai&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">eachoverlap</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="s">&quot;Chr2&quot;</span><span class="p">,</span> <span class="mi">10000</span><span class="o">:</span><span class="mi">11000</span><span class="p">)</span>
    <span class="c"># `record` is a BAM.Record object</span>
    <span class="c"># ...</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><code>eachoverlap</code> also accepts the <code>Interval</code> type defined in <a href="https://github.com/BioJulia/GenomicFeatures.jl">GenomicFeatures.jl</a>.</p>
<div class="codehilite"><pre><span></span><span class="c"># Load GFF3 module.</span>
<span class="k">using</span> <span class="n">GenomicFeatures</span>

<span class="c"># Load genomic features from a GFF3 file.</span>
<span class="n">features</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">collect</span><span class="p">,</span> <span class="n">GFF3</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;TAIR10_GFF3_genes.gff&quot;</span><span class="p">)</span>

<span class="c"># Keep mRNA features.</span>
<span class="n">filter!</span><span class="p">(</span><span class="n">x</span> <span class="o">-&gt;</span> <span class="n">GFF3</span><span class="o">.</span><span class="n">featuretype</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="s">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">features</span><span class="p">)</span>

<span class="c"># Load BAM module.</span>
<span class="k">using</span> <span class="n">BioAlignments</span>

<span class="c"># Open a BAM file and iterate over records overlapping mRNA transcripts.</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;SRR1238088.sort.bam&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s">&quot;SRR1238088.sort.bam.bai&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">feature</span> <span class="kp">in</span> <span class="n">features</span>
    <span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">eachoverlap</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span>
        <span class="c"># `record` overlaps `feature`.</span>
        <span class="c"># ...</span>
    <span class="k">end</span>
<span class="k">end</span>
<span class="n">close</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
</pre></div>


<p><a id='Performance-tips-1'></a></p>
<h2 id="performance-tips">Performance tips</h2>
<p>The size of a BAM file is often extremely large. The iterator interface mentioned above allocates an object for each record and that may be a bottleneck of reading data from a BAM file. In-place reading reuses a preallocated object for every record and less memory allocation happens in reading:</p>
<div class="codehilite"><pre><span></span><span class="n">reader</span> <span class="o">=</span> <span class="n">open</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">BAM</span><span class="o">.</span><span class="n">Record</span><span class="p">()</span>
<span class="k">while</span> <span class="o">!</span><span class="n">eof</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
    <span class="n">read!</span><span class="p">(</span><span class="n">reader</span><span class="p">,</span> <span class="n">record</span><span class="p">)</span>
    <span class="c"># do something</span>
<span class="k">end</span>
</pre></div>


<p>Accessing optional fields will results in type instability in Julia, which has a significant negative impact on performance. If the user knows the type of a value in advance, specifying it as a type annotation will alleviate the problem:</p>
<div class="codehilite"><pre><span></span><span class="k">for</span> <span class="n">record</span> <span class="kp">in</span> <span class="n">open</span><span class="p">(</span><span class="n">BAM</span><span class="o">.</span><span class="n">Reader</span><span class="p">,</span> <span class="s">&quot;data.bam&quot;</span><span class="p">)</span>
    <span class="n">nm</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;NM&quot;</span><span class="p">]</span><span class="o">::</span><span class="kt">UInt8</span>
    <span class="c"># do something</span>
<span class="k">end</span>
</pre></div>
                
                  
                
              
              
                
              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../pairalign/" title="Pairwise alignment" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Pairwise alignment
              </span>
            </div>
          </a>
        
        
          <a href="../references/" title="References" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                References
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright (c) 2017 The BioJulia Organisation
          </div>
        
        powered by
        <a href="http://www.mkdocs.org" title="MkDocs">MkDocs</a>
        and
        <a href="http://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application-1d6ee6ee6c.js"></script>
      
      
      <script>app.initialize({url:{base:".."}})</script>
      
        <script src="../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>